src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PopupControlService.cs

```diff
@@ -1173,19 +1137,400 @@ internal KeyboardFocusChangedEventHandler FocusChangedEventHandler
+                        // with rightmost will increase for a while, and the first decrease
+                        // occurs at the desired maximum.
+                        int maxIndex = prevRightmostIndex;
+                        for (; maxIndex < hullCount; ++maxIndex)
+                        {
+                            int wrapIndex = maxIndex + 1;
+                            if (wrapIndex == hullCount) wrapIndex = 0;
+                            if (Cross(rightmost, points[maxIndex], points[wrapIndex]) < 0)
+                                break;
+                        }
+
+                        // replace the segment of the hull between these two points with
+                        // the leftmost and rightmost point(s)
+                        int pointsToRemove = maxIndex - minIndex - 1;
+                        int delta = pointsToAdd - pointsToRemove;
+
+                        // move retained points to their new position
+                        // (the hull is a subset of the original points, which
+                        // guarantees that the indices into points are
+                        // always in bounds).
+                        if (delta < 0)
+                        {
+                            for (int i=maxIndex; i<hullCount; ++i)
+                            {
+                                points[i + delta] = points[i];
+                            }
+                        }
+                        else if (delta > 0)
+                        {
+                            for (int i=hullCount-1; i>=maxIndex; --i)
+                            {
+                                points[i + delta] = points[i];
+                            }
+                        }
+
+                        // insert the new point(s), and update the hull size
+                        points[minIndex + 1] = leftmost;
+                        prevLeftmostIndex = prevRightmostIndex = minIndex + 1;
+                        if (pointsToAdd == 2)
+                        {
+                            points[minIndex + 2] = rightmost;
+                            prevRightmostIndex = minIndex + 2;
+                        }
+                        hullCount += delta;
+                    }
+                }
+
+                // when the loop terminates, the loop invariant plus the condition
+                // (currentIndex >= N) imply points[0 ... hullCount-1] describe the
+                // convex hull of the original points.  All that's left is to discard
+                // any extra points, and compute the directions.
+                points.RemoveRange(hullCount, N - hullCount);
+                _points = points.ToArray();
+                SetDirections();
+            }
+
+            // set the Direction field on each point.  This enables optimizations during
+            // ContainsPoint.
+            private void SetDirections()
+            {
+                for (int i=0, N=_points.Length; i<N; ++i)
```

Comment:
Should we set the `_points[0].Direction = Direction.Skew;` when the `_points.Length` is one?

Reply:
That can't happen in our use-case - the convex hull of a non-zero number of non-empty rectangles will have at least 4 points.  In general, the convex hull of any set of points will have at least 3 points unless it's 1-dimensional (all points are collinear, hull has 2 points) or 0-dimensional (the set is a single point, hull has 1 point).  

Our ContainsPoint implementation only works when the hull has 2-d area (which is the only case we'll see).  Changing the direction to Skew in the 1-point case doesn't fix that - it will still give the wrong answer, just for a different reason.